### PR TITLE
[3.16] Add support for codeAction/resolve

### LINKED ISSUE #2469: [3.16] Support CodeActionsResolve request
[SPEC](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#codeAction_resolve)

* Add client capabilities `textDocument.codeAction.resolveSupport`
* :leftwards_arrow_with_hook: `codeAction/resolve`

This will increase performance if the server supports it

### PR DESCRIPTION
**Pull request by ericdallo**
Fixes #2469

### CODE REVIEW

File: lsp-mode.el [hunk 1]
```diff
@@ -5103,10 +5106,24 @@ It will show up only if current point has signature help."
...
+(defun lsp-execute-code-action (action)
   "Execute code action ACTION.
-If ACTION is not set it will be selected from `lsp-code-actions-at-point'."
+If ACTION is not set it will be selected from `lsp-code-actions-at-point'.
+Request codeAction/resolve for more info if server supports."
   (interactive (list (lsp--select-action (lsp-code-actions-at-point))))
+  (if (-> (lsp--server-capabilities)
+          (lsp:server-capabilities-code-action-provider?)
+          (lsp:code-action-options-resolve-provider?))
+      (lsp-request-async
```
Comment (yyoncho) [inline lsp-mode.el:5117]: I would expect that to be sync. Imagine that you perform the code action and you switch the buffer - the code action won't be performed.

File: lsp-mode.el [hunk 2]
```diff
@@ -5103,10 +5106,24 @@ It will show up only if current point has signature help."
...
     (lsp--send-execute-command command arguments?)))
 
-(lsp-defun lsp-execute-code-action ((action &as &CodeAction :command? :edit?))
+(defun lsp-execute-code-action (action)
   "Execute code action ACTION.
-If ACTION is not set it will be selected from `lsp-code-actions-at-point'."
+If ACTION is not set it will be selected from `lsp-code-actions-at-point'.
+Request codeAction/resolve for more info if server supports."
   (interactive (list (lsp--select-action (lsp-code-actions-at-point))))
+  (if (-> (lsp--server-capabilities)
```
Comment (yyoncho) [inline lsp-mode.el:5114]: please use `lsp-feature?` for that(refer to `lsp-method-requirements`).

Comment (yyoncho) [inline lsp-mode.el:5114]: shouldn't we perform resolve only if the property we are looking for is not present. This is how resolve works for code lenses if memory serves me right.

File: lsp-mode.el [hunk 3]
```diff
@@ -3200,7 +3200,9 @@ disappearing, unset all the variables related to it."
                                                                                                   "refactor.inline"
                                                                                                   "refactor.rewrite"
                                                                                                   "source"
-                                                                                                  "source.organizeImports"])))))))
+                                                                                                  "source.organizeImports"])))))
+                                     (resolveSupport . ((properties . ["edit" "command" "kind"])))
```
Comment (yyoncho) [inline lsp-mode.el:3204]: hm, we perform actions like filtering based on "kind". In that case, it seems to me that it is better to always request kind. WDYT?

File: lsp-mode.el (hunk 1 above)
Comment (ericdallo) [reply lsp-mode.el:5117]: I thought this is expected since we wouldn't block user input, but I'll change it

File: lsp-mode.el (hunk 2 above)
Comment (ericdallo) [reply lsp-mode.el:5114]: I searched about it but for server capabilities, I could not find how to do that, do you know how I can check for the resolveSupport via that?

Comment (ericdallo) [reply lsp-mode.el:5114]: Yeah, I thought about that too but I was not sure, in this case we should check if `command` or `workspaceEdit` is present, right?

File: lsp-mode.el (hunk 3 above)
Comment (ericdallo) [reply lsp-mode.el:3204]: Hum, I think it makes sense, I'll remove it from here

File: lsp-mode.el (hunk 2 above)
Comment (yyoncho) [reply lsp-mode.el:5114]: I think it is edit not workspaceEdit but yes.

Comment (yyoncho) [reply lsp-mode.el]: Check how textDocument/prepareRename is handled in lsp-method-requirements.

Comment (nbfalcon) [reply lsp-mode.el]: You'll need to use `(lsp-feature? "codeAction/resolve")` and to add `("codeAction/resolve" :capability :resolveSupport)` to `lsp-method-requirements`.

Comment (ericdallo) [reply lsp-mode.el]: Nice, I'll change it

Comment (yyoncho) [reply lsp-mode.el]: > You'll need to use `(lsp-feature? "codeAction/resolve")` and to add `("codeAction/resolve" :capability :resolveSupport)` to `lsp-method-requirements`.

resolveProvider is not top level, so a custom check-command will be needed.

Comment (ericdallo) [reply lsp-mode.el]: Done!

**Review by yyoncho — APPROVED** [review]

**Pull request merged**